homeassistant/components/text/__init__.py

```diff
@@ -0,0 +1,194 @@
+        )
+    if entity.pattern is not None and re.compile(entity.pattern).match(value) is None:
+        raise ValueError(
+            f"Value {value} for {entity.name} doesn't match pattern {entity.pattern}"
+        )
+    await entity.async_set_value(value)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up a config entry."""
+    component: EntityComponent[TextEntity] = hass.data[DOMAIN]
+    return await component.async_setup_entry(entry)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    component: EntityComponent[TextEntity] = hass.data[DOMAIN]
+    return await component.async_unload_entry(entry)
+
+
+class TextMode(StrEnum):
+    """Modes for text entities."""
+
+    TEXT = "text"
+    PASSWORD = "password"
+
+
+@dataclass
+class TextEntityDescription(EntityDescription):
+    """A class that describes text entities."""
+
+
+class TextEntity(Entity):
+    """Representation of a Text entity."""
+
+    entity_description: TextEntityDescription
+    _attr_mode: TextMode = TextMode.TEXT
+    _attr_native_value: str | None
+    _attr_native_min: int = 0
+    _attr_native_max: int = 100
+    _attr_pattern: str | None = None
+    _attr_min: None = None
+    _attr_max: None = None
+    _attr_state: None = None
+    _attr_value: None = None
+
+    @property
+    def capability_attributes(self) -> dict[str, Any]:
+        """Return capability attributes."""
+        return {
+            ATTR_MODE: self.mode,
+            ATTR_MIN: self.min,
+            ATTR_MAX: self.max,
+            ATTR_PATTERN: self.pattern,
+        }
+
+    @property
+    @final
+    def state(self) -> str | None:
+        """Return the entity state."""
+        return self.value
+
+    @property
+    def mode(self) -> TextMode:
+        """Return the mode of the entity."""
+        return self._attr_mode
+
+    @property
+    def native_min(self) -> int:
+        """Return the minimum length of the value."""
+        return self._attr_native_min
+
+    @property
+    @final
+    def min(self) -> int:
+        """Return the minimum length of the value."""
+        return max(self.native_min, 0)
+
+    @property
+    def native_max(self) -> int:
+        """Return the maximum length of the value."""
+        return self._attr_native_max
+
+    @property
+    @final
+    def max(self) -> int:
+        """Return the maximum length of the value."""
+        return min(self.native_max, MAX_LENGTH_STATE_STATE)
+
+    @property
+    def pattern(self) -> str | None:
+        """Return the regex pattern that the value must match."""
+        return self._attr_pattern
+
+    @property
+    def native_value(self) -> str | None:
+        """Return the value reported by the text."""
+        return self._attr_native_value
+
+    @property
+    @final
+    def value(self) -> str | None:
```

Comment:
This can just be the `state` method, right? The existing state property is now a proxy, I suggest to merge them.